languages/elixir/exercises/concept/bitstrings/.docs/after.md

```diff
@@ -0,0 +1,84 @@
+## Bitstrings
+
+- [Bitstrings][bitstring] allow programmers to work with binary data effectively.
+
+  - Can construct [bitstrings][bitstring] elegantly using the `<<>>` special form
+
+  ```elixir
+  <<0b101::3>> == <<5::3>> and <<5::3>> == <<5::size(3)>>
+  # => true
+
+  first = <<0b110::3>>
+  second = <<0b001::3>>
+  combined = <<first::bitstring, second::bitstring>>
+  # => <<49::size(6)>>
+  ```
+
+  - Can match using the `<<>>` form and use that match to further decompose binary data
+
+  ```elixir
+  <<data_size::32, count::size(data_size)>> = <<8::32, 3::8>>
+  # => <<0, 0, 0, 8, 3>>
+  data_size
+  # => 8
+  count
+  # => 3
+  ```
+
+  - Or you can match on the remainder of the [bitstring][bitstring] using `::bitstring`
+
+  ```elixir
+  <<_::32, remainder::bitstring>> = <<8::32, 3::8>>
+  # => <<0, 0, 0, 8, 3>>
+  remainder
+  # => <<3>>
+  ```
+
+- The binary data*type* is a specific type of [bitstring][bitstring].
+  - All binaries are [bitstrings][bitstring], but not all [bitstrings][bitstring] are binaries.
+- Elixir provides syntactic sugar for working with [integer literals][integer-literal] in different forms:
+  - `0b1000001 = ?A = 65`
+
+## Tail Call Recursion and Accumulators
+
+- If a language has [tail call optimization][recursion-tco], it often mitigates concerns of overflowing the function call stack.
+- In order to achieve [tail call recursion][recursion-tco] in Elixir, an accumulator is often required because of data immutability rules.
+
+  - An accumulator is variable that is passed along the recursive call keeping the current state of the recursive function _so far_ until the final result is returned in the base case.
+
+  ```elixir
+  # Count the length of a list without an accumulator
+  def count([]), do: 0
+  def count([_head | tail]), do: 1 + count(tail)
+
+  # Count the length of a list with an accumulator
+  def count(list), do: do_count(list, 0)
+
+  defp do_count([], count), do: count
+  defp do_count([_head | tail], count), do: do_count(tail, count + 1)
+  ```
+
+- When working with lists, remember they are implemented as singly-linked lists: prepend operations are very fast, but append operations are very slow.
+
+  ```elixir
+  # Reverse a list with an accumulator
+  def reverse(list), do: do_reverse(list)
```

Comment:
I have been thinking about the pros and cons of initializing the accumulator when the function with the accumulator is first being called versus as a default argument. 
```elixir
def reverse(list), do: do_reverse(list, [])

# vs

def reverse(list), do: do_reverse(list)
defp do_reverse(list, acc \\ [])
```

If we assume that the private function is not reused in multiple public function, then I only see downsides in the default argument approach: it has one extra line, and it makes the public function look like it doesn't do anything (of course it does, it hides the possibility to choose the accumulator from the outside, but it's not immediately obvious).

What do you think?

Reply:
I agree, that the benefit is mixed, which is why I wanted to show it two ways:
  - count in the introduction does not use a default argument
  - reverse in the after does use a default argument

The thing I think is a benefit of the header even in a private function is that it implies the intended contract (in the absence of typespecs).  Of course as you said, it could be seen as redundant because of the public scoped function.  I have no strong feeling about this and your points I think fall on the side of safety when teaching idioms -- I will remove the header.